docs/history-service.md

```diff
@@ -0,0 +1,274 @@
+```
+
+<details>
+<summary><i>Code entrypoints</i></summary>
+
+These state transitions are performed using the generic utility function [GetAndUpdateWorkflowWithNew](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/api/update_workflow_util.go#L37). The utility function receives a function that performs the updates to mutable state specific to the particular state transition, and then performs the two commits to persistence via
+[`UpdateWorkflowExecutionAsActive`](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/workflow/context.go#L459), ultimately calling the persistence backends, e.g.
+[sql](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/common/persistence/sql/execution.go#L339) and [cassandra](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/common/persistence/cassandra/execution_store.go#L142).
+
+- When handling a `ScheduleActivityTask` command from a Worker, a task is [added to in-memory mutable state](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/workflow/task_generator.go#L419) representing the Transfer Task that must be persisted.
+
+Call sites:
+
+- Handling an RPC from the User Application, for example [`SignalWorkflow`](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/api/signalworkflow/api.go#L54).
+
+- Handling a `WorkflowTaskCompleted` RPC from the Worker in [`handleWorkflowTaskCompleted`](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/workflow_task_handler_callbacks.go#L690)
+
+- Timer task processor: [`executeUserTimerTimeoutTask`](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/timer_queue_active_task_executor.go#L136) handles a `workflow.sleep()` timer set in user code by appending `TimerFired`, and `WorkflowTaskScheduled` events to Workflow History, and creating a Transfer Task for the Workflow Task.
+
+</details>
+<br>
+<br>
+
+#### Consistency guarantees
+
+- Consistency of Mutable State and History Tasks is achieved via use of database transactions.
+- Consistency of these with History Events is achieved by storing in Mutable State the identity of the latest History Event reflected in Mutable State; a History Event is only "valid" if it is in Mutable State and, on failure to persist dirty Mutable State and associated History Tasks, we reload from persistence.
+- Consistency with Matching Service is achieved by the combination of the above within-History Service consistency guarantees, and the Transfer Task queue processor which ensures that Transfer Tasks written by a History Shard will eventually result in the required Workflow or Activity Task being created in the Matching Service (this is known as the "Transactional Outbox" Pattern).
```

Comment:
it might be good to talk about the other parts of the history↔matching protocol here. i.e. if matching can't dispatch a task immediately (sync), it has to store it durably in its own persistence before returning from the add task rpc. and matching has to get a successful response from a recordxtaskstarted rpc to history before dispatching a task to a worker.

Reply:
I think I'd like to leave that to a subsequent update of this document, or for the (currently missing) `docs/matching-service.md`, and quite possibly by a different author. (I'm already at risk of over-reaching myself here.)